addons/purchase_stock/models/product.py

```diff
@@ -69,3 +69,49 @@ def _get_quantity_in_progress(self, location_ids=False, warehouse_ids=False):
             qty_by_product_location[(product.id, location.id)] += product_qty
             qty_by_product_wh[(product.id, location.warehouse_id.id)] += product_qty
         return qty_by_product_location, qty_by_product_wh
+
+
+class SupplierInfo(models.Model):
+    _inherit = 'product.supplierinfo'
+
+    last_purchase_date = fields.Date('Last Purchase', compute='_compute_last_purchase_date')
+    show_set_supplier_button = fields.Boolean(
+        'Show Set Supplier Button', compute='_compute_show_set_supplier_button')
+
+    def _compute_last_purchase_date(self):
+        self.last_purchase_date = False
+        purchases = self.env['purchase.order'].search([
```

Comment:
if you have lot of purchase from one product (for exemple 1 million), it can make lot of time to search.
Make one query per supplier info
```python
for supplier in self:
      supplier.last_purchase_date = self.env['purchase.order'].sudo().search([('state', 'in', ('purchase', 'done')),
            ('order_line.product_id', 'in',
             supplier.product_tmpl_id.product_variant_ids.ids),
            ('partner_id', '=', supplier.name.id),
        ], order='date_order', limit=1)..date_order
```

@amoyaux

Reply:
ping @ryv-odoo